feat(AIP-8): clarify clients AIPs should benefit

toumorokoshi opened the pull request:
Without clear guidance on the types of clients AIPs should benefit, it is unclear to an AIP author what the criteria of a good AIP is.

This clarifies the guidance, and implicitly expand the scope of AIPs beneficiaries from just end-users to all clients that user may use to interact with the API.

toumorokoshi committed:
feat(AIP-8): clarify clients AIPs should benefit

Without clear guidance on the types of clients AIPs should
benefit, it is unclear to an AIP author what the criteria of
a good AIP is.

This clarifies the guidance, and implicitly expand the
scope of AIPs beneficiaries from just end-users to all
clients that user may use to interact with the API.

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,40 @@ placement:
-Google and the way API teams discuss and come to consensus on API guidance.
+providing high-level, concise documentation for API design and development. The
+goal is for these documents to serve as the source of truth for API-related
+documentation at Google and the way API teams discuss and come to consensus on
```
I see this is pre-existing, but is this part of the AIPs specific to Google?

toumorokoshi replied:
yes. The general AIPs actually say something a bit different: https://aip-dev.github.io/aip.dev/1.

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,40 @@ placement:
+  followed.
+
+AIPs **must** contain API guidance that provide a clear benefit to one or more
+clients, including but not limited to:
```
types of clients

bgrant0607 replied:
Or:
to authors of one or more types of clients, ...

toumorokoshi replied:
I think it's fair to say that an AIP could benefit a client directly (e.g. reducing the complexity of the interface exposed) or an author (include more annotations to reduce the need to hand-annotate).

So I adjusted to include clients *or* their authors.

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,40 @@ placement:
+- [Infrastructure as Code][] tools.
+- SDKs to interact with an API from a programming language.
+- Visual User Interfaces.
+- Users.
```
IMO, "users" is vague and is what led to the previous style-guide approach.

toumorokoshi replied:
Users is indeed vague, and I think that if guidance was *only* suppose to benefit the user, then it leaves a lot of room for interpretation.

However, this section codifies that *in addition to* users, one must consider the clients as well. I think it's important to consider users in AIPs since there is guidance that's tangible to them (e.g. consistency in units and naming)

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,40 @@ placement:
+- Command line interfaces.
+- [Infrastructure as Code][] tools.
+- SDKs to interact with an API from a programming language.
+- Visual User Interfaces.
```
More types, to expand peoples' imaginations a bit:
* asset inventory tools and/or operation on resource data at rest
* Security Orchestration, Automation, and Response / Remediation
* recommendation tools
* controllers
* test frameworks
* simple clients (e.g., scripts)

toumorokoshi replied:
awesome list! added.

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,40 @@ placement:
+- Visual User Interfaces.
+- Users.
+
+AIP guidance **must not** contain guidance that significantly detriments a
```
AFAIK, "detriment" is a noun

toumorokoshi replied:
ack, rephrased.

toumorokoshi replied:
ptal.

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -191,12 +209,32 @@ subdirectory); this ensures that the link works both on the AIP site, when
+purpose and target beneficiary.
+
+The beneficiaries of improved API design are users. These users interact with
+APIs via a variety of methods, depending on their use case:
```
via a variety of clients, since that's the topic here

"methods" could also be interpreted as rpcs

toumorokoshi replied:
ack, good call.

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -191,12 +209,32 @@ subdirectory); this ensures that the link works both on the AIP site, when
+- Command line interfaces for exploration and simple automation.
+- [IaC][] clients for orchestration and automation of multiple resources.
+- Visual User Interfaces for visualization and one-off manual actions.
+- SDKs for data-plane operations.
```
or custom controllers

toumorokoshi replied:
ack, added.

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -191,12 +209,32 @@ subdirectory); this ensures that the link works both on the AIP site, when
+- Visual User Interfaces for visualization and one-off manual actions.
+- SDKs for data-plane operations.
+
+API guidance must in turn consider each of these different clients.
```
I'd say a "broad range of types of clients", to tie it back to the subsection header.

toumorokoshi replied:
I went with "these clients" to refer to the above list. WDYT?

toumorokoshi replied:
actually I did some wordsmithing and moved the list to joint it with the more comprehensive list in the guidance. PTAL.

toumorokoshi committed:
addressing feedback

- consolidating the list of clients to a single location.
- adding more lists of clients.

toumorokoshi committed:
Merge branch 'master' into toum/clarify-clients

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,49 @@ placement:
+  guidance **must** clearly explain under what conditions the guidance should be
+  followed.
+
+AIPs **must** contain API guidance that provide a clear benefit to authors of
```
This sentence now contains redundant text: "to authors of
one or more types of clients or their authors". Maybe just remove "or their authors".

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,49 @@ placement:
+- Custom controllers (e.g. auto-scalers) which poll live state and adjust
+  resource configruation accordingly.
+- [IaC][] clients for orchestration and automation of multiple resources.
+- Reccomendation tools which provide guidance on which APIs are useful for
```
Typo: Reccomendation

bgrant0607 commented on aip/general/0008.md:
```diff
@@ -7,22 +7,49 @@ placement:
+- Visual User Interfaces for visualization and one-off manual actions.
+- Users.
+
+AIP guidance **must not** be a signifcant detriment to a client's usability.
```
Typo: signifcant

bgrant0607 replied:
Also, this mentions usability, but we also care about implementation difficulty.

toumorokoshi replied:
ack, added implementation difficulty.

bgrant0607 reviewed (changes requested):
A few typos and one significant comment, regarding implementation difficulty.

toumorokoshi committed:
addressing feedback

- fixing typos and grammar

toumorokoshi committed:
Merge branch 'master' into toum/clarify-clients

bgrant0607 reviewed (approved):
Thanks for the updates.

noahdietz commented on aip/general/0008.md:
```diff
@@ -7,22 +7,50 @@ placement:
+  followed.
+
+AIPs **must** contain API guidance that provide a clear benefit to one or more
+types of clients or their authors, including but not limited to:
```
Can you include an example (existing or otherwise) of what this looks like? Feels a bit nebulous to me how this will manifest in an AIP. Is this a new subsection in Rationale? I'd like to be able to point at something that fulfills this requirement, rather than debate what content does or does not address it.

toumorokoshi replied:
sorry for the delay. I added examples, can you see if that helps?

noahdietz replied:
Ok thanks. Last clarification, is this guidance (that you are adding with the `**must**`) to say that an AIP **must** spell out how it is beneficial to (a) clients? Or is this more of a meta point to say that "if the proposed guidance is cumbersome in these ways, it won't be accepted"?

Basically, how is this guidance here:
`AIPs **must** contain API guidance that provide a clear`

actionable in AIP text if at all

toumorokoshi replied:
I read it as "an AIP must provide a clear benefit to a client or user". I wasn't intending to even say it has to spell it out. 

It's the same level of ambiguity as "cover a single, discrete topic": that's somewhat subjective, but maybe clear enough to most that one can action on it (e.g. you say "split this up into separate AIPs").

In the case of this wording, the action would be "it's not clear that this guidance improves clients or users".

But yeah this could be read differently... how about "must contain API guidance that benefits one or more"... just getting rid of the "clear" part. 

What do you think? open to ideas too, assuming my intention is clear and you agree.

noahdietz replied:
I see what you mean! I see a requirement word and I automatically think about how the guidance would manifest concretely, so might just be me.

Idk why but the following reads better to me, but I'm not picky and will approve this as is:
```md
Guidance contained within an AIP **must** be beneficial to one or more
types of clients or their authors, including but not limited to:
```

toumorokoshi replied:
updated! thanks for the brainstorming.

toumorokoshi committed:
Merge branch 'master' into toum/clarify-clients

toumorokoshi committed:
Merge branch 'master' into toum/clarify-clients

toumorokoshi committed:
addressing feedback

toumorokoshi committed:
Merge branch 'master' into toum/clarify-clients

toumorokoshi committed:
addressing feedback
